migrations/000032_app_versions_table.up.sql

```diff
@@ -0,0 +1,489 @@
+BEGIN;
+
+SET search_path = public, pg_catalog;
+
+--
+-- Drop obsolete views that will not work with versions.
+--
+DROP VIEW IF EXISTS app_job_types;
+DROP VIEW IF EXISTS vice_analyses;
+
+--
+-- Drop views that depend on columns which will be dropped,
+-- or that will add new view columns.
+--
+DROP VIEW IF EXISTS tool_listing;
+DROP VIEW IF EXISTS app_listing;
+DROP VIEW IF EXISTS job_listings;
+
+--
+-- app_versions table
+--
+CREATE TABLE IF NOT EXISTS app_versions (
+    id uuid NOT NULL DEFAULT uuid_generate_v1(),
+    app_id uuid NOT NULL,
+    version varchar(255) NOT NULL,
+    version_order int NOT NULL DEFAULT 0,
+    deleted boolean NOT NULL DEFAULT false,
+    disabled boolean NOT NULL DEFAULT false,
+    integration_data_id uuid NOT NULL,
+    integration_date timestamp without time zone,
+    edited_date timestamp without time zone DEFAULT CURRENT_TIMESTAMP,
+
+    PRIMARY KEY (id),
+    FOREIGN KEY (app_id) REFERENCES apps(id) ON DELETE CASCADE,
+    UNIQUE (app_id, version_order)
+);
+
+CREATE INDEX IF NOT EXISTS app_versions_app_id ON app_versions(app_id);
+CREATE INDEX IF NOT EXISTS app_versions_version_order ON app_versions(version_order);
+
+--
+-- Initial app_versions data
+--
+INSERT INTO app_versions (
+    app_id,
+    version,
+    deleted,
+    disabled,
+    integration_data_id,
+    integration_date,
+    edited_date
+)
+    SELECT
+        id AS app_id,
+        'Unversioned' AS version,
+        deleted,
+        disabled,
+        integration_data_id,
+        integration_date,
+        edited_date
+    FROM apps
+    ON CONFLICT DO NOTHING;
+
+--
+-- Replace the `app_id` column in the `app_steps` table
+-- with a new `app_version_id` column.
+--
+ALTER TABLE ONLY app_steps
+    ADD COLUMN IF NOT EXISTS app_version_id uuid;
+UPDATE app_steps
+    SET app_version_id = (
+        SELECT id
+        FROM app_versions
+        WHERE app_steps.app_id = app_versions.app_id
+    );
+ALTER TABLE ONLY app_steps
+    ALTER COLUMN app_version_id SET NOT NULL;
+ALTER TABLE ONLY app_steps
+    DROP COLUMN IF EXISTS app_id;
+
+--
+-- Foreign Key for `app_version_id` column in `app_steps` table.
+--
+ALTER TABLE ONLY app_steps
+    ADD CONSTRAINT app_steps_app_version_id_fkey
+    FOREIGN KEY (app_version_id)
+    REFERENCES app_versions(id) ON DELETE CASCADE;
+
+CREATE INDEX IF NOT EXISTS app_steps_app_version_id ON app_steps(app_version_id);
+
+--
+-- Add `app_version_id` column to the `jobs` table.
+--
+ALTER TABLE ONLY jobs
+    ADD COLUMN IF NOT EXISTS app_version_id uuid;
+UPDATE jobs
+    SET app_version_id = (
+        SELECT id
+        FROM app_versions
+        WHERE jobs.app_id = CAST(app_versions.app_id AS character varying)
+    );
```

Comment:
It looks like the `app_id` column is remaining in the `jobs` table? It might be fine, but we should maybe be prepared for it going out of sync since we're denormalizing a little bit there.

Reply:
I'll check again, but I'm pretty sure that the `app_id` column in the `jobs` table can also refer to TAPIS app IDs (which are not UUIDs).

Reply:
That's correct; it can refer to TAPIS app IDs as well.

Reply:
ah, of course, we don't have an external vs. internal ID split for apps. so we've already got the potential for denormalization issues, and this doesn't change anything about that.